File: Oc-Dome/Oc-Dome/Classes/Tools(工具)/HWCollectionViewController/HWCollectionViewController.m
//
//  HWCollectionViewController.m
//  Oc-Dome
//
//  Created by 李含文 on 2018/12/1.
//  Copyright © 2018年 东莞市三心网络科技有限公司. All rights reserved.
//

#import "HWCollectionViewController.h"


@interface HWCollectionViewController ()

@property(nonatomic, copy) UICollectionViewCell *(^cellForRow)(NSIndexPath *indexPath);
@property(nonatomic, copy)  CGSize(^cellSize)(NSIndexPath *indexPath);
@property(nonatomic, copy) UICollectionReusableView *(^reusableView)(NSString *kind,NSIndexPath *indexPath);
@property(nonatomic, copy) CGSize(^headerSize)(NSInteger section);
@property(nonatomic, copy) CGSize (^footerSize)(NSInteger section);

@property(nonatomic, assign) BOOL isAutoRefresh;

@end

@implementation HWCollectionViewController
- (void)setPage:(NSInteger)page {
    _page = page;
}
- (void)setFlowLayout:(UICollectionViewLayout *)flowLayout {
    _flowLayout = flowLayout;
    [self.collectionView.collectionViewLayout invalidateLayout];
    [self.collectionView setCollectionViewLayout:flowLayout];
}
- (NSMutableArray *)itemArray {
    if (!_itemArray) {
        _itemArray = [NSMutableArray array];
    }
    return _itemArray;
}
- (UICollectionView *)collectionView {
    if (!_collectionView) {
        UICollectionViewLayout *fl;
        if(self.flowLayout) {
            fl = self.flowLayout;
        } else {
            UICollectionViewFlowLayout *myfl = [[UICollectionViewFlowLayout alloc] init];
            myfl.itemSize = CGSizeMake([UIScreen mainScreen].bounds.size.width, 50);
            myfl.minimumLineSpacing = 0;
            myfl.minimumInteritemSpacing = 0;
            myfl.scrollDirection = UICollectionViewScrollDirectionVertical;
            fl = myfl;
        }
        CGRect frame = CGRectMake(0, self.view.y, HWScreenW, HWScreenH - hw_navHeight);
        _collectionView = [[UICollectionView alloc] initWithFrame:frame collectionViewLayout:fl];
        _collectionView.delegate = self;
        _collectionView.dataSource = self;
        _collectionView.backgroundColor = [UIColor clearColor];
        __weak typeof(self) weakSelf = self;
        _collectionView.hw_ishasDataBlock = ^(BOOL tag) {
            weakSelf.noDataView.hidden = tag;
        };
    }
    return _collectionView;
}
- (void)viewDidLoad {
    [super viewDidLoad];
    self.automaticallyAdjustsScrollViewInsets = NO;
}
- (void)addNoDataView {
    self.noDataView = [HWNoDataView noDataView];
    [self.collectionView addSubview:self.noDataView];
    [self.collectionView sendSubviewToBack:self.noDataView];
}
- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    if (!self.collectionView.mj_header.isRefreshing && self.isAutoRefresh == true) {
        self.page = 1;
        [self loadData:self.page];
    }
}

- (void)loadData:(NSInteger)page {}

- (void)setRefreshType:(HWRefreshType)type isAuto:(BOOL)isAuto {
    if (![self.view.subviews containsObject:self.collectionView]) {return;}
    [self addNoDataView];
    self.isAutoRefresh = isAuto;
    __weak typeof(self) weakSelf = self;
    switch (type) {
        case HWRefreshTypeDefault: {
            self.collectionView.hw_hearderRefreshBlock = ^{
                weakSelf.page = 1;
                [weakSelf loadData:weakSelf.page];
            };
            self.collectionView.hw_footerRefreshBlock = ^{
                weakSelf.page += 1;
                [weakSelf loadData:weakSelf.page];
            };
            [self.collectionView beginRefreshing];
            break;}
        case HWRefreshTypeHearder: {
            self.collectionView.hw_hearderRefreshBlock = ^{
                weakSelf.page = 1;
                [weakSelf loadData:weakSelf.page];
            };
            [self.collectionView beginRefreshing];
            break;}
        case HWRefreshTypeFooter: {
            self.page = 1;
            [weakSelf loadData:weakSelf.page];
            self.collectionView.hw_footerRefreshBlock = ^{
                weakSelf.page += 1;
                [weakSelf loadData:weakSelf.page];
            };
            break;}
        default:
            break;
    }
    if (self.collectionView.mj_footer) {
        [self.collectionView addObserver:self forKeyPath:@"contentSize" options:NSKeyValueObservingOptionNew context:nil];
    }
}
- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    if ([keyPath isEqualToString:@"contentSize"]) {
        if (!self.collectionView.mj_footer) {return;}
        CGFloat offset = self.collectionView.contentSize.height;
        if (offset < self.collectionView.height - hw_navHeight) {
            self.collectionView.mj_footer.hidden = YES;
        } else {
            self.collectionView.mj_footer.hidden = NO;
        }
    }
}
- (void)endRefreshing {
    if (![self.view.subviews containsObject:self.collectionView]) {return;}
    if (!self.collectionView.mj_header && !self.collectionView.mj_footer) {return;};
    [self.collectionView endRefreshing];
}
- (void)beginRefreshing {
    if (![self.view.subviews containsObject:self.collectionView]) {return;}
    if (!self.collectionView.mj_header) {return;};
    [self.collectionView beginRefreshing];
}
- (void)setCellForRowBlock:(UICollectionViewCell *(^)(NSIndexPath *indexPath))cellForRowBlock
             cellSizeBlock:(CGSize(^)(NSIndexPath *indexPath))cellSizeBlock {
    [self.view addSubview:self.collectionView];
    self.cellForRow = cellForRowBlock;
    self.cellSize = cellSizeBlock;
}
- (void)setReusableViewBlock:(UICollectionReusableView *(^)(NSString *kind,NSIndexPath *indexPath))reusableViewBlock
             headerSizeBlock:(CGSize(^)(NSInteger section))headerSizeBlock
             footerSizeBlock:(CGSize(^)(NSInteger section))footerSizeBlock {
    if (![self.view.subviews containsObject:self.collectionView]) {return;}
    self.reusableView = reusableViewBlock;
    self.headerSize = headerSizeBlock;
    self.footerSize = footerSizeBlock;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.itemArray.count;
}
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    return self.cellForRow(indexPath);
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    return self.cellSize(indexPath);
}
- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath {
    return self.reusableView(kind,indexPath);
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout referenceSizeForHeaderInSection:(NSInteger)section {
    if (self.headerSize == nil){
        return CGSizeZero;
    }
    return self.headerSize(section);
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout referenceSizeForFooterInSection:(NSInteger)section {
    if (self.footerSize == nil){
        return CGSizeZero;
    }
    return self.footerSize(section);
}
- (void)dealloc {
    if (_collectionView.mj_footer) {
        [_collectionView removeObserver:self forKeyPath:@"contentSize"];
    }
}

@end

File: Swift-Dome/LHWDome/Classes/Tool(工具类)/New Group/HWCollectionViewController.swift
//
//  HWCollectionViewController.swift
//  LHWDome
//
//  Created by 李含文 on 2019/7/8.
//  Copyright © 2019年 李含文. All rights reserved.
//

import UIKit
import SnapKit
import EmptyDataSet_Swift

enum HWRefreshType : Int { // 刷新类型
    case `default` = 0 // 默认没有
    case dropDown = 1  // 只支持下拉
    case all = 2       // 支持上拉和下拉
}

class HWCollectionViewController: BaseViewController {

    /// 内部模型数组
    open lazy var itemArray = NSMutableArray()
    /// 角标
    open var page = 1
    /// 触发刷新就会来
    open func loadData() {}
    /// 刷新样式
    open var type = HWRefreshType.default {
        willSet {
           setRefresh(newValue)
        }
    }
    open var setCollectionViewFrameBlock:((UICollectionView)->())?
    /// 从新设置布局参数
    open var fl: UICollectionViewLayout? {
        willSet {
            guard newValue != nil else {
                return
            }
            self.collectionView.collectionViewLayout.invalidateLayout()
            collectionView.setCollectionViewLayout((newValue)!, animated: true)
        }
    }
    open lazy var collectionView : UICollectionView = {
        let fl = UICollectionViewFlowLayout()
        // myflowLayout.sectionHeadersPinToVisibleBounds = true // 头部悬浮
        fl.minimumLineSpacing = 0
        fl.minimumInteritemSpacing = 0
        let collectionView = UICollectionView(frame:self.view.bounds, collectionViewLayout: fl)
        collectionView.dataSource = self
        collectionView.delegate = self
        collectionView.backgroundColor = UIColor.clear
        collectionView.emptyDataSetSource = self
        collectionView.emptyDataSetDelegate = self
        return collectionView
    }()
    override func viewDidLoad() {
        super.viewDidLoad()
        if #available(iOS 11.0, *) {
            collectionView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentBehavior.never
        } else {
            automaticallyAdjustsScrollViewInsets = false
        }
        self.view.addSubview(self.collectionView)
    }
//    override func viewDidLayoutSubviews() {
//        if setCollectionViewFrameBlock == nil {
//            collectionView.snp.makeConstraints { (make) in
//                make.edges.equalToSuperview()
//            }
//        }
//    }
    /// 默认flowLayout样式
    open func setCollectionView(cellForItem: @escaping ((_ indexPath:IndexPath) -> UICollectionViewCell),
                           cellSize: @escaping ((_ indexPath:IndexPath) -> CGSize)) {
        cellForItemBlock = cellForItem
        cellSizeBlock = cellSize
    }
    open func setReusableView(headerViewSize: @escaping ((_ section : Int) -> CGSize),
                              footerViewSize: @escaping ((_ section : Int) -> CGSize),
                              reusableView: @escaping ((_ indexPath:IndexPath, _ kind: String) -> UICollectionReusableView)) {
        headerViewSizeBlock = headerViewSize
        footerViewSizeBlock = footerViewSize
        reusableViewBlock = reusableView
    }
    // MARK: - 闭包
    private var cellForItemBlock : ((_ index:IndexPath) -> UICollectionViewCell)?
    private var cellSizeBlock : ((_ index:IndexPath) -> CGSize)?
    private var headerViewSizeBlock : ((_ section : Int) -> CGSize)?
    private var footerViewSizeBlock : ((_ section : Int) -> CGSize)?
    private var reusableViewBlock : ((_ index:IndexPath, _ kind: String) -> UICollectionReusableView)?

    private func setRefresh(_ type: HWRefreshType) {
        switch type {
        case .default:
            collectionView.mj_header = nil
            collectionView.mj_footer = nil
        case .dropDown:
            collectionView.hw_hearderRefreshBlock = {
                self.page = 1
                self.loadData()
            }
        default:
            collectionView.hw_hearderRefreshBlock = {
                self.page = 1
                self.loadData()
            }
            collectionView.hw_footerRefreshBlock = {
                self.loadData()
            }
        }
    }
}
extension HWCollectionViewController: UICollectionViewDataSource {
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.itemArray.count
    }
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        return self.cellForItemBlock?(indexPath) ?? UICollectionViewCell()
    }
    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        return self.reusableViewBlock?(indexPath, kind) ?? UICollectionReusableView()
    }
}
extension HWCollectionViewController : UICollectionViewDelegateFlowLayout{
    /// 设置cell CGSize
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return self.cellSizeBlock?(indexPath) ?? CGSize.zero
    }
    /// 设置headerView CGSize
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForHeaderInSection section: Int) -> CGSize {
        return self.headerViewSizeBlock?(section) ?? CGSize.zero
    }
    /// 设置footerView CGSize
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForFooterInSection section: Int) -> CGSize {
        return self.footerViewSizeBlock?(section) ?? CGSize.zero
    }
}

// MARK: 隐藏上拉刷新
extension HWCollectionViewController : UIScrollViewDelegate {
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        let size = scrollView.contentSize
        if size.height < collectionView.height{
            self.collectionView.mj_footer?.isHidden = true
        } else {
            self.collectionView.mj_footer?.isHidden = false
        }
    }
}

extension HWCollectionViewController: EmptyDataSetSource, EmptyDataSetDelegate {
    func image(forEmptyDataSet scrollView: UIScrollView) -> UIImage? {
        return hw_Image(named: "没数据缺省页")
    }
    func title(forEmptyDataSet scrollView: UIScrollView) -> NSAttributedString? {
        return "暂无数据".hw_toAttribute().hw_color(UIColor.lightGray).hw_fontSize(15)
    }
    func verticalOffset(forEmptyDataSet scrollView: UIScrollView) -> CGFloat {
        return -70
    }
    func emptyDataSet(_ scrollView: UIScrollView, didTapView view: UIView) {
        HWPrint("点击了屏幕")
    }
}
